File: AllModels.swift
import Foundation

// ============= RESIDENT MODEL =============
struct Resident: Codable, Identifiable {
    let id: String
    let name: String
    let email: String
    let phone: String
    let unitNumber: String
    let status: ResidentStatus
    let createdAt: Date
    let updatedAt: Date
    let tenantId: String
    let profileImageURL: String?
    let emergencyContact: EmergencyContact?
    let preferences: ResidentPreferences
    
    enum ResidentStatus: String, Codable, CaseIterable {
        case active = "active"
        case inactive = "inactive"
        case pending = "pending"
        case suspended = "suspended"
        
        var displayText: String {
            switch self {
            case .active: return "Activo"
            case .inactive: return "Inactivo"
            case .pending: return "Pendiente"
            case .suspended: return "Suspendido"
            }
        }
        
        var color: UIColor {
            switch self {
            case .active: return .systemGreen
            case .inactive: return .systemGray
            case .pending: return .systemOrange
            case .suspended: return .systemRed
            }
        }
    }
    
    init(id: String, name: String, email: String, phone: String, unitNumber: String, status: ResidentStatus) {
        self.id = id
        self.name = name
        self.email = email
        self.phone = phone
        self.unitNumber = unitNumber
        self.status = status
        self.createdAt = Date()
        self.updatedAt = Date()
        self.tenantId = ""
        self.profileImageURL = nil
        self.emergencyContact = nil
        self.preferences = ResidentPreferences()
    }
}

struct EmergencyContact: Codable {
    let name: String
    let phone: String
    let relationship: String
}

struct ResidentPreferences: Codable {
    let language: String
    let notifications: NotificationPreferences
    let privacy: PrivacySettings
    
    init() {
        self.language = "es"
        self.notifications = NotificationPreferences()
        self.privacy = PrivacySettings()
    }
}

struct NotificationPreferences: Codable {
    let email: Bool
    let sms: Bool
    let push: Bool
    let announcements: Bool
    let maintenance: Bool
    let payments: Bool
    let security: Bool
    
    init() {
        self.email = true
        self.sms = true
        self.push = true
        self.announcements = true
        self.maintenance = true
        self.payments = true
        self.security = true
    }
}

struct PrivacySettings: Codable {
    let shareContactInfo: Bool
    let allowDirectMessages: Bool
    let showInDirectory: Bool
    
    init() {
        self.shareContactInfo = false
        self.allowDirectMessages = true
        self.showInDirectory = true
    }
}

// ============= PAYMENT MODEL =============
struct Payment: Codable, Identifiable {
    let id: String
    let amount: Double
    let method: String
    let status: PaymentStatus
    let date: Date
    let description: String?
    let transactionId: String?
    let residentId: String?
    
    enum PaymentStatus: String, Codable, CaseIterable {
        case pending = "pending"
        case processing = "processing"
        case completed = "completed"
        case failed = "failed"
        case refunded = "refunded"
        
        var displayText: String {
            switch self {
            case .pending: return "Pendiente"
            case .processing: return "Procesando"
            case .completed: return "Completado"
            case .failed: return "Fallido"
            case .refunded: return "Reembolsado"
            }
        }
        
        var color: UIColor {
            switch self {
            case .pending: return .systemOrange
            case .processing: return .systemBlue
            case .completed: return .systemGreen
            case .failed: return .systemRed
            case .refunded: return .systemPurple
            }
        }
    }
    
    init(id: String, amount: Double, method: String, status: PaymentStatus, date: Date) {
        self.id = id
        self.amount = amount
        self.method = method
        self.status = status
        self.date = date
        self.description = nil
        self.transactionId = nil
        self.residentId = nil
    }
}

// ============= MAINTENANCE MODEL =============
struct WorkOrder: Codable, Identifiable {
    let id: String
    let title: String
    let description: String
    let status: WorkOrderStatus
    let priority: Priority
    let type: String
    let location: String
    let createdAt: Date
    let scheduledDate: Date?
    let completedDate: Date?
    let assignedTo: String?
    let estimatedCost: Double?
    let actualCost: Double?
    
    enum WorkOrderStatus: String, Codable, CaseIterable {
        case pending = "pending"
        case assigned = "assigned"
        case inProgress = "in_progress"
        case completed = "completed"
        case cancelled = "cancelled"
        
        var displayText: String {
            switch self {
            case .pending: return "Pendiente"
            case .assigned: return "Asignado"
            case .inProgress: return "En Progreso"
            case .completed: return "Completado"
            case .cancelled: return "Cancelado"
            }
        }
        
        var color: UIColor {
            switch self {
            case .pending: return .systemOrange
            case .assigned: return .systemBlue
            case .inProgress: return .systemYellow
            case .completed: return .systemGreen
            case .cancelled: return .systemRed
            }
        }
    }
    
    enum Priority: String, Codable, CaseIterable {
        case low = "low"
        case medium = "medium"
        case high = "high"
        case critical = "critical"
        
        var displayText: String {
            switch self {
            case .low: return "Baja"
            case .medium: return "Media"
            case .high: return "Alta"
            case .critical: return "Crítica"
            }
        }
        
        var color: UIColor {
            switch self {
            case .low: return .systemGreen
            case .medium: return .systemYellow
            case .high: return .systemOrange
            case .critical: return .systemRed
            }
        }
    }
    
    init(id: String, title: String, description: String, status: WorkOrderStatus, priority: Priority, type: String, location: String, createdAt: Date, scheduledDate: Date?) {
        self.id = id
        self.title = title
        self.description = description
        self.status = status
        self.priority = priority
        self.type = type
        self.location = location
        self.createdAt = createdAt
        self.scheduledDate = scheduledDate
        self.completedDate = nil
        self.assignedTo = nil
        self.estimatedCost = nil
        self.actualCost = nil
    }
}

// ============= SECURITY MODEL =============
struct SecurityEvent: Codable, Identifiable {
    let id: String
    let type: String
    let description: String
    let severity: Severity
    let location: String
    let timestamp: Date
    let status: String?
    let userId: String?
    let deviceId: String?
    
    enum Severity: String, Codable, CaseIterable {
        case info = "info"
        case low = "low"
        case medium = "medium"
        case high = "high"
        case critical = "critical"
        
        var displayText: String {
            switch self {
            case .info: return "Información"
            case .low: return "Baja"
            case .medium: return "Media"
            case .high: return "Alta"
            case .critical: return "Crítica"
            }
        }
        
        var color: UIColor {
            switch self {
            case .info: return .systemBlue
            case .low: return .systemGreen
            case .medium: return .systemYellow
            case .high: return .systemOrange
            case .critical: return .systemRed
            }
        }
    }
    
    init(id: String, type: String, description: String, severity: Severity, location: String, timestamp: Date) {
        self.id = id
        self.type = type
        self.description = description
        self.severity = severity
        self.location = location
        self.timestamp = timestamp
        self.status = nil
        self.userId = nil
        self.deviceId = nil
    }
}

struct Visitor: Codable, Identifiable {
    let id: String
    let name: String
    let phone: String
    let visitingUnit: String
    let visitDate: Date
    let visitTime: String
    let status: VisitorStatus
    let purpose: String
    let authorizedBy: String?
    let checkInTime: Date?
    let checkOutTime: Date?
    
    enum VisitorStatus: String, Codable, CaseIterable {
        case expected = "expected"
        case checkedIn = "checked_in"
        case checkedOut = "checked_out"
        case denied = "denied"
        case expired = "expired"
        
        var displayText: String {
            switch self {
            case .expected: return "Esperado"
            case .checkedIn: return "Ingresado"
            case .checkedOut: return "Salió"
            case .denied: return "Denegado"
            case .expired: return "Expirado"
            }
        }
        
        var color: UIColor {
            switch self {
            case .expected: return .systemBlue
            case .checkedIn: return .systemGreen
            case .checkedOut: return .systemGray
            case .denied: return .systemRed
            case .expired: return .systemOrange
            }
        }
    }
    
    init(id: String, name: String, phone: String, visitingUnit: String, visitDate: Date, visitTime: String, status: VisitorStatus, purpose: String) {
        self.id = id
        self.name = name
        self.phone = phone
        self.visitingUnit = visitingUnit
        self.visitDate = visitDate
        self.visitTime = visitTime
        self.status = status
        self.purpose = purpose
        self.authorizedBy = nil
        self.checkInTime = nil
        self.checkOutTime = nil
    }
}

// ============= COMMUNICATION MODEL =============
struct Message: Codable, Identifiable {
    let id: String
    let senderId: String
    let senderName: String
    let recipientId: String
    let subject: String
    let content: String
    let timestamp: Date
    let isRead: Bool
    let messageType: MessageType
    let attachments: [Attachment]?
    
    enum MessageType: String, Codable {
        case personal = "personal"
        case announcement = "announcement"
        case system = "system"
        case emergency = "emergency"
    }
}

struct Announcement: Codable, Identifiable {
    let id: String
    let title: String
    let content: String
    let author: String
    let timestamp: Date
    let priority: Priority
    let expirationDate: Date?
    let targetAudience: [String]?
    
    enum Priority: String, Codable {
        case low = "low"
        case medium = "medium"
        case high = "high"
        case urgent = "urgent"
    }
}

struct Attachment: Codable {
    let id: String
    let filename: String
    let fileType: String
    let fileSize: Int
    let url: String
}

// ============= REPORT MODEL =============
struct Report: Codable, Identifiable {
    let id: String
    let name: String
    let type: ReportType
    let status: ReportStatus
    let createdAt: Date
    let completedAt: Date?
    let fileUrl: String?
    let parameters: [String: String]?
    
    enum ReportType: String, Codable, CaseIterable {
        case financial = "financial"
        case operational = "operational"
        case maintenance = "maintenance"
        case security = "security"
        case residents = "residents"
        
        var displayText: String {
            switch self {
            case .financial: return "Financiero"
            case .operational: return "Operacional"
            case .maintenance: return "Mantenimiento"
            case .security: return "Seguridad"
            case .residents: return "Residentes"
            }
        }
    }
    
    enum ReportStatus: String, Codable {
        case generating = "generating"
        case completed = "completed"
        case failed = "failed"
        
        var displayText: String {
            switch self {
            case .generating: return "Generando"
            case .completed: return "Completado"
            case .failed: return "Fallido"
            }
        }
    }
}

// ============= USER MODEL =============
struct User: Codable, Identifiable {
    let id: String
    let username: String
    let email: String
    let name: String
    let phone: String?
    let role: UserRole
    let tenantId: String
    let isActive: Bool
    let lastLogin: Date?
    let createdAt: Date
    let preferences: UserPreferences?
    
    enum UserRole: String, Codable, CaseIterable {
        case superAdmin = "super_admin"
        case admin = "admin"
        case propertyManager = "property_manager"
        case resident = "resident"
        case maintenance = "maintenance"
        case security = "security"
        
        var displayText: String {
            switch self {
            case .superAdmin: return "Super Administrador"
            case .admin: return "Administrador"
            case .propertyManager: return "Gerente de Propiedad"
            case .resident: return "Residente"
            case .maintenance: return "Mantenimiento"
            case .security: return "Seguridad"
            }
        }
    }
}

struct UserPreferences: Codable {
    let language: String
    let theme: String
    let notifications: NotificationPreferences
    let privacy: PrivacySettings
}

// ============= CONFIGURATION MODEL =============
struct Configuration: Codable {
    let tenantId: String
    let settings: [String: Any]
    let features: [String: Bool]
    let integrations: [String: IntegrationConfig]
    let updatedAt: Date
    
    private enum CodingKeys: String, CodingKey {
        case tenantId, settings, features, integrations, updatedAt
    }
}

struct IntegrationConfig: Codable {
    let enabled: Bool
    let apiKey: String?
    let endpoint: String?
    let settings: [String: String]?
}

// ============= ACTIVITY MODEL =============
struct Activity: Codable, Identifiable {
    let id: String
    let type: ActivityType
    let description: String
    let timestamp: Date
    let userId: String?
    let entityId: String?
    let entityType: String?
    
    enum ActivityType: String, Codable {
        case newResident = "new_resident"
        case payment = "payment"
        case maintenance = "maintenance"
        case security = "security"
        case communication = "communication"
        case system = "system"
    }
}

// ============= TASK MODEL =============
struct Task: Codable, Identifiable {
    let id: String
    let title: String
    let description: String?
    let priority: Priority
    let status: TaskStatus
    let assignedTo: String?
    let createdBy: String
    let dueDate: Date?
    let completedDate: Date?
    let createdAt: Date
    
    enum Priority: String, Codable {
        case low = "low"
        case medium = "medium"
        case high = "high"
        
        var displayText: String {
            switch self {
            case .low: return "Baja"
            case .medium: return "Media"
            case .high: return "Alta"
            }
        }
    }
    
    enum TaskStatus: String, Codable {
        case pending = "pending"
        case inProgress = "in_progress"
        case completed = "completed"
        case cancelled = "cancelled"
    }
    
    init(id: String, title: String, priority: Priority, dueDate: Date?) {
        self.id = id
        self.title = title
        self.description = nil
        self.priority = priority
        self.status = .pending
        self.assignedTo = nil
        self.createdBy = ""
        self.dueDate = dueDate
        self.completedDate = nil
        self.createdAt = Date()
    }
}

// ============= API RESPONSE MODELS =============
struct APIResponse<T: Codable>: Codable {
    let success: Bool
    let data: T?
    let message: String?
    let error: String?
}

struct PaginatedResponse<T: Codable>: Codable {
    let items: [T]
    let totalCount: Int
    let page: Int
    let pageSize: Int
    let hasMore: Bool
}

File: AllModels.js
// ============= RESIDENT MODEL =============
class Resident {
    constructor(data) {
        this.id = data.id;
        this.name = data.name;
        this.email = data.email;
        this.phone = data.phone;
        this.unitNumber = data.unitNumber;
        this.status = data.status || 'active';
        this.emergencyContact = data.emergencyContact || {};
        this.preferences = data.preferences || {};
        this.tenantId = data.tenantId;
        this.createdAt = data.createdAt || new Date().toISOString();
        this.updatedAt = data.updatedAt || new Date().toISOString();
    }

    static validate(data) {
        const errors = [];
        
        if (!data.name || data.name.trim().length < 2) {
            errors.push('Name must be at least 2 characters long');
        }
        
        if (!data.email || !this.isValidEmail(data.email)) {
            errors.push('Valid email is required');
        }
        
        if (!data.phone || !this.isValidPhone(data.phone)) {
            errors.push('Valid phone number is required');
        }
        
        if (!data.unitNumber || data.unitNumber.trim().length === 0) {
            errors.push('Unit number is required');
        }
        
        if (data.status && !['active', 'inactive', 'pending', 'suspended'].includes(data.status)) {
            errors.push('Invalid status value');
        }
        
        return errors;
    }

    static isValidEmail(email) {
        const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
        return emailRegex.test(email);
    }

    static isValidPhone(phone) {
        const phoneRegex = /^\+?[\d\s\-\(\)]{10,}$/;
        return phoneRegex.test(phone);
    }

    toJSON() {
        return {
            id: this.id,
            name: this.name,
            email: this.email,
            phone: this.phone,
            unitNumber: this.unitNumber,
            status: this.status,
            emergencyContact: this.emergencyContact,
            preferences: this.preferences,
            tenantId: this.tenantId,
            createdAt: this.createdAt,
            updatedAt: this.updatedAt
        };
    }
}

// ============= PAYMENT MODEL =============
class Payment {
    constructor(data) {
        this.id = data.id;
        this.residentId = data.residentId;
        this.residentName = data.residentName;
        this.residentEmail = data.residentEmail;
        this.unitNumber = data.unitNumber;
        this.amount = data.amount;
        this.method = data.method;
        this.description = data.description || '';
        this.status = data.status || 'pending';
        this.date = data.date || new Date().toISOString();
        this.transactionId = data.transactionId;
        this.refundInfo = data.refundInfo || null;
        this.tenantId = data.tenantId;
        this.createdAt = data.createdAt || new Date().toISOString();
        this.updatedAt = data.updatedAt || new Date().toISOString();
    }

    static validate(data) {
        const errors = [];
        
        if (!data.residentId) {
            errors.push('Resident ID is required');
        }
        
        if (!data.amount || data.amount <= 0) {
            errors.push('Amount must be greater than 0');
        }
        
        if (!data.method || !['credit_card', 'debit_card', 'bank_transfer', 'cash', 'check'].includes(data.method)) {
            errors.push('Valid payment method is required');
        }
        
        if (data.status && !['pending', 'processing', 'completed', 'failed', 'refunded'].includes(data.status)) {
            errors.push('Invalid status value');
        }
        
        return errors;
    }

    calculateFees() {
        const feeRates = {
            'credit_card': 0.029, // 2.9%
            'debit_card': 0.015,  // 1.5%
            'bank_transfer': 0.005, // 0.5%
            'cash': 0,
            'check': 0
        };
        
        const rate = feeRates[this.method] || 0;
        return Math.round(this.amount * rate * 100) / 100;
    }

    getNetAmount() {
        return this.amount - this.calculateFees();
    }

    canRefund() {
        return this.status === 'completed' && !this.refundInfo;
    }

    toJSON() {
        return {
            id: this.id,
            residentId: this.residentId,
            residentName: this.residentName,
            residentEmail: this.residentEmail,
            unitNumber: this.unitNumber,
            amount: this.amount,
            method: this.method,
            description: this.description,
            status: this.status,
            date: this.date,
            transactionId: this.transactionId,
            refundInfo: this.refundInfo,
            fees: this.calculateFees(),
            netAmount: this.getNetAmount(),
            tenantId: this.tenantId,
            createdAt: this.createdAt,
            updatedAt: this.updatedAt
        };
    }
}

// ============= WORK ORDER MODEL =============
class WorkOrder {
    constructor(data) {
        this.id = data.id;
        this.title = data.title;
        this.description = data.description;
        this.status = data.status || 'pending';
        this.priority = data.priority;
        this.type = data.type;
        this.location = data.location;
        this.assignedTo = data.assignedTo || null;
        this.scheduledDate = data.scheduledDate || null;
        this.completedDate = data.completedDate || null;
        this.estimatedCost = data.estimatedCost || null;
        this.actualCost = data.actualCost || null;
        this.tenantId = data.tenantId;
        this.createdBy = data.createdBy;
        this.createdAt = data.createdAt || new Date().toISOString();
        this.updatedAt = data.updatedAt || new Date().toISOString();
    }

    static validate(data) {
        const errors = [];
        
        if (!data.title || data.title.trim().length < 3) {
            errors.push('Title must be at least 3 characters long');
        }
        
        if (!data.description || data.description.trim().length < 10) {
            errors.push('Description must be at least 10 characters long');
        }
        
        if (!data.priority || !['low', 'medium', 'high', 'critical'].includes(data.priority)) {
            errors.push('Valid priority is required');
        }
        
        if (!data.type || !['plumbing', 'electrical', 'hvac', 'cleaning', 'painting', 'landscaping', 'security', 'other'].includes(data.type)) {
            errors.push('Valid type is required');
        }
        
        if (!data.location || data.location.trim().length === 0) {
            errors.push('Location is required');
        }
        
        if (data.status && !['pending', 'assigned', 'in_progress', 'completed', 'cancelled'].includes(data.status)) {
            errors.push('Invalid status value');
        }
        
        if (data.estimatedCost && data.estimatedCost < 0) {
            errors.push('Estimated cost cannot be negative');
        }
        
        if (data.actualCost && data.actualCost < 0) {
            errors.push('Actual cost cannot be negative');
        }
        
        return errors;
    }

    isOverdue() {
        if (!this.scheduledDate || this.status === 'completed' || this.status === 'cancelled') {
            return false;
        }
        
        return new Date(this.scheduledDate) < new Date();
    }

    getDaysUntilDue() {
        if (!this.scheduledDate) return null;
        
        const now = new Date();
        const dueDate = new Date(this.scheduledDate);
        const diffTime = dueDate - now;
        const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
        
        return diffDays;
    }

    getPriorityScore() {
        const priorityScores = {
            'low': 1,
            'medium': 2,
            'high': 3,
            'critical': 4
        };
        
        return priorityScores[this.priority] || 0;
    }

    toJSON() {
        return {
            id: this.id,
            title: this.title,
            description: this.description,
            status: this.status,
            priority: this.priority,
            type: this.type,
            location: this.location,
            assignedTo: this.assignedTo,
            scheduledDate: this.scheduledDate,
            completedDate: this.completedDate,
            estimatedCost: this.estimatedCost,
            actualCost: this.actualCost,
            isOverdue: this.isOverdue(),
            daysUntilDue: this.getDaysUntilDue(),
            priorityScore: this.getPriorityScore(),
            tenantId: this.tenantId,
            createdBy: this.createdBy,
            createdAt: this.createdAt,
            updatedAt: this.updatedAt
        };
    }
}

// ============= SECURITY EVENT MODEL =============
class SecurityEvent {
    constructor(data) {
        this.id = data.id;
        this.type = data.type;
        this.description = data.description;
        this.location = data.location;
        this.severity = data.severity;
        this.timestamp = data.timestamp || new Date().toISOString();
        this.userId = data.userId || null;
        this.userName = data.userName || null;
        this.resolved = data.resolved || false;
        this.resolvedAt = data.resolvedAt || null;
        this.resolvedBy = data.resolvedBy || null;
        this.tenantId = data.tenantId;
    }

    static validate(data) {
        const errors = [];
        
        if (!data.type || !['access_granted', 'access_denied', 'alarm_triggered', 'motion_detected', 'door_opened', 'camera_offline', 'system_error', 'emergency_alert'].includes(data.type)) {
            errors.push('Valid event type is required');
        }
        
        if (!data.description || data.description.trim().length < 5) {
            errors.push('Description must be at least 5 characters long');
        }
        
        if (!data.location || data.location.trim().length === 0) {
            errors.push('Location is required');
        }
        
        if (!data.severity || !['info', 'low', 'medium', 'high', 'critical'].includes(data.severity)) {
            errors.push('Valid severity level is required');
        }
        
        return errors;
    }

    isCritical() {
        return this.severity === 'critical' || this.severity === 'high';
    }

    getAgeInMinutes() {
        const now = new Date();
        const eventTime = new Date(this.timestamp);
        return Math.floor((now - eventTime) / (1000 * 60));
    }

    resolve(resolvedBy) {
        this.resolved = true;
        this.resolvedAt = new Date().toISOString();
        this.resolvedBy = resolvedBy;
    }

    toJSON() {
        return {
            id: this.id,
            type: this.type,
            description: this.description,
            location: this.location,
            severity: this.severity,
            timestamp: this.timestamp,
            userId: this.userId,
            userName: this.userName,
            resolved: this.resolved,
            resolvedAt: this.resolvedAt,
            resolvedBy: this.resolvedBy,
            isCritical: this.isCritical(),
            ageInMinutes: this.getAgeInMinutes(),
            tenantId: this.tenantId
        };
    }
}

// ============= VISITOR MODEL =============
class Visitor {
    constructor(data) {
        this.id = data.id;
        this.name = data.name;
        this.phone = data.phone;
        this.visitingUnit = data.visitingUnit;
        this.visitDate = data.visitDate;
        this.visitTime = data.visitTime;
        this.purpose = data.purpose;
        this.status = data.status || 'expected';
        this.checkInTime = data.checkInTime || null;
        this.checkOutTime = data.checkOutTime || null;
        this.photo = data.photo || null;
        this.tenantId = data.tenantId;
        this.createdBy = data.createdBy;
        this.createdAt = data.createdAt || new Date().toISOString();
    }

    static validate(data) {
        const errors = [];
        
        if (!data.name || data.name.trim().length < 2) {
            errors.push('Name must be at least 2 characters long');
        }
        
        if (!data.phone || !Resident.isValidPhone(data.phone)) {
            errors.push('Valid phone number is required');
        }
        
        if (!data.visitingUnit || data.visitingUnit.trim().length === 0) {
            errors.push('Visiting unit is required');
        }
        
        if (!data.visitDate) {
            errors.push('Visit date is required');
        }
        
        if (!data.visitTime) {
            errors.push('Visit time is required');
        }
        
        if (!data.purpose || data.purpose.trim().length < 3) {
            errors.push('Purpose must be at least 3 characters long');
        }
        
        if (data.status && !['expected', 'checked_in', 'checked_out', 'denied', 'expired'].includes(data.status)) {
            errors.push('Invalid status value');
        }
        
        return errors;
    }

    isExpired() {
        if (this.status === 'checked_out' || this.status === 'denied') {
            return false;
        }
        
        const visitDateTime = new Date(`${this.visitDate} ${this.visitTime}`);
        const expirationTime = new Date(visitDateTime.getTime() + (24 * 60 * 60 * 1000)); // 24 hours
        
        return new Date() > expirationTime;
    }

    getDurationInMinutes() {
        if (!this.checkInTime || !this.checkOutTime) {
            return null;
        }
        
        const checkIn = new Date(this.checkInTime);
        const checkOut = new Date(this.checkOutTime);
        
        return Math.floor((checkOut - checkIn) / (1000 * 60));
    }

    checkIn() {
        if (this.status !== 'expected') {
            throw new Error('Visitor cannot be checked in');
        }
        
        this.status = 'checked_in';
        this.checkInTime = new Date().toISOString();
    }

    checkOut() {
        if (this.status !== 'checked_in') {
            throw new Error('Visitor cannot be checked out');
        }
        
        this.status = 'checked_out';
        this.checkOutTime = new Date().toISOString();
    }

    toJSON() {
        return {
            id: this.id,
            name: this.name,
            phone: this.phone,
            visitingUnit: this.visitingUnit,
            visitDate: this.visitDate,
            visitTime: this.visitTime,
            purpose: this.purpose,
            status: this.status,
            checkInTime: this.checkInTime,
            checkOutTime: this.checkOutTime,
            photo: this.photo,
            isExpired: this.isExpired(),
            durationInMinutes: this.getDurationInMinutes(),
            tenantId: this.tenantId,
            createdBy: this.createdBy,
            createdAt: this.createdAt
        };
    }
}

// ============= MESSAGE MODEL =============
class Message {
    constructor(data) {
        this.id = data.id;
        this.senderUserId = data.senderUserId;
        this.senderName = data.senderName;
        this.recipientId = data.recipientId;
        this.subject = data.subject;
        this.content = data.content;
        this.messageType = data.messageType || 'personal';
        this.priority = data.priority || 'medium';
        this.isRead = data.isRead || false;
        this.readAt = data.readAt || null;
        this.attachments = data.attachments || [];
        this.timestamp = data.timestamp || new Date().toISOString();
        this.tenantId = data.tenantId;
    }

    static validate(data) {
        const errors = [];
        
        if (!data.senderUserId) {
            errors.push('Sender user ID is required');
        }
        
        if (!data.recipientId) {
            errors.push('Recipient ID is required');
        }
        
        if (!data.subject || data.subject.trim().length < 3) {
            errors.push('Subject must be at least 3 characters long');
        }
        
        if (!data.content || data.content.trim().length < 5) {
            errors.push('Content must be at least 5 characters long');
        }
        
        if (data.messageType && !['personal', 'announcement', 'system', 'emergency'].includes(data.messageType)) {
            errors.push('Invalid message type');
        }
        
        if (data.priority && !['low', 'medium', 'high', 'urgent'].includes(data.priority)) {
            errors.push('Invalid priority level');
        }
        
        return errors;
    }

    markAsRead() {
        this.isRead = true;
        this.readAt = new Date().toISOString();
    }

    getAgeInHours() {
        const now = new Date();
        const messageTime = new Date(this.timestamp);
        return Math.floor((now - messageTime) / (1000 * 60 * 60));
    }

    isUrgent() {
        return this.priority === 'urgent' || this.messageType === 'emergency';
    }

    toJSON() {
        return {
            id: this.id,
            senderUserId: this.senderUserId,
            senderName: this.senderName,
            recipientId: this.recipientId,
            subject: this.subject,
            content: this.content,
            messageType: this.messageType,
            priority: this.priority,
            isRead: this.isRead,
            readAt: this.readAt,
            attachments: this.attachments,
            timestamp: this.timestamp,
            ageInHours: this.getAgeInHours(),
            isUrgent: this.isUrgent(),
            tenantId: this.tenantId
        };
    }
}

// ============= ANNOUNCEMENT MODEL =============
class Announcement {
    constructor(data) {
        this.id = data.id;
        this.title = data.title;
        this.content = data.content;
        this.priority = data.priority || 'medium';
        this.authorId = data.authorId;
        this.author = data.author;
        this.expirationDate = data.expirationDate || null;
        this.timestamp = data.timestamp || new Date().toISOString();
        this.tenantId = data.tenantId;
    }

    static validate(data) {
        const errors = [];
        
        if (!data.title || data.title.trim().length < 5) {
            errors.push('Title must be at least 5 characters long');
        }
        
        if (!data.content || data.content.trim().length < 10) {
            errors.push('Content must be at least 10 characters long');
        }
        
        if (!data.authorId) {
            errors.push('Author ID is required');
        }
        
        if (data.priority && !['low', 'medium', 'high', 'urgent'].includes(data.priority)) {
            errors.push('Invalid priority level');
        }
        
        if (data.expirationDate && new Date(data.expirationDate) <= new Date()) {
            errors.push('Expiration date must be in the future');
        }
        
        return errors;
    }

    isExpired() {
        if (!this.expirationDate) return false;
        return new Date(this.expirationDate) <= new Date();
    }

    isActive() {
        return !this.isExpired();
    }

    getDaysUntilExpiration() {
        if (!this.expirationDate) return null;
        
        const now = new Date();
        const expiration = new Date(this.expirationDate);
        const diffTime = expiration - now;
        const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
        
        return diffDays > 0 ? diffDays : 0;
    }

    toJSON() {
        return {
            id: this.id,
            title: this.title,
            content: this.content,
            priority: this.priority,
            authorId: this.authorId,
            author: this.author,
            expirationDate: this.expirationDate,
            timestamp: this.timestamp,
            isExpired: this.isExpired(),
            isActive: this.isActive(),
            daysUntilExpiration: this.getDaysUntilExpiration(),
            tenantId: this.tenantId
        };
    }
}

module.exports = {
    Resident,
    Payment,
    WorkOrder,
    SecurityEvent,
    Visitor,
    Message,
    Announcement
};
